// @hashgraph/sdk/lib/account/TransferTransaction.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Hbar = _interopRequireDefault(require("../Hbar.cjs"));

var _TokenId = _interopRequireDefault(require("../token/TokenId.cjs"));

var _AccountId = _interopRequireDefault(require("./AccountId.cjs"));

var _Transaction = _interopRequireWildcard(require("../transaction/Transaction.cjs"));

var _long = _interopRequireDefault(require("long"));

var _NullableTokenDecimalMap = _interopRequireDefault(require("./NullableTokenDecimalMap.cjs"));

var _TokenTransferMap = _interopRequireDefault(require("./TokenTransferMap.cjs"));

var _HbarTransferMap = _interopRequireDefault(require("./HbarTransferMap.cjs"));

var _TokenNftTransferMap = _interopRequireDefault(require("./TokenNftTransferMap.cjs"));

var util = _interopRequireWildcard(require("../util.cjs"));

var _NftId = _interopRequireDefault(require("../token/NftId.cjs"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @typedef {import("../long.js").LongObject} LongObject
 * @typedef {import("bignumber.js").default} BigNumber
 */

/**
 * @namespace proto
 * @typedef {import("@hashgraph/proto").ITransaction} proto.ITransaction
 * @typedef {import("@hashgraph/proto").ISignedTransaction} proto.ISignedTransaction
 * @typedef {import("@hashgraph/proto").TransactionBody} proto.TransactionBody
 * @typedef {import("@hashgraph/proto").ITransactionBody} proto.ITransactionBody
 * @typedef {import("@hashgraph/proto").ITransactionResponse} proto.ITransactionResponse
 * @typedef {import("@hashgraph/proto").ICryptoTransferTransactionBody} proto.ICryptoTransferTransactionBody
 * @typedef {import("@hashgraph/proto").ITokenID} proto.ITokenID
 * @typedef {import("@hashgraph/proto").IAccountID} proto.IAccountID
 * @typedef {import("@hashgraph/proto").IAccountAmount} proto.IAccountAmount
 * @typedef {import("@hashgraph/proto").ITokenTransferList} proto.ITokenTransferList
 */

/**
 * @typedef {import("../channel/Channel.js").default} Channel
 * @typedef {import("../client/Client.js").default<*, *>} Client
 * @typedef {import("../transaction/TransactionId.js").default} TransactionId
 */

/**
 * @typedef {object} TransferTokensInput
 * @property {TokenId | string} tokenId
 * @property {AccountId | string} accountId
 * @property {Long | number} amount
 */

/**
 * @typedef {object} TransferTokenObject
 * @property {TokenId} tokenId
 * @property {AccountId} accountId
 * @property {Long} amount
 */

/**
 * @typedef {object} TransferHbarInput
 * @property {AccountId | string} accountId
 * @property {number | string | Long | BigNumber | Hbar} amount
 */

/**
 * @typedef {object} TransferNftInput
 * @property {TokenId | string} tokenId
 * @property {AccountId | string} sender
 * @property {AccountId | string} recipient
 * @property {Long | number} serial
 */

/**
 * Transfers a new Hedera™ crypto-currency token.
 */
class TransferTransaction extends _Transaction.default {
  /**
   * @param {object} [props]
   * @param {(TransferTokensInput)[]} [props.tokenTransfers]
   * @param {(TransferHbarInput)[]} [props.hbarTransfers]
   * @param {(TransferNftInput)[]} [props.nftTransfers]
   */
  constructor(props = {}) {
    super();
    /**
     * @private
     * @type {TokenTransferMap}
     */

    this._tokenTransfers = new _TokenTransferMap.default();
    /**
     * @private
     * @type {NullableTokenDecimalMap}
     */

    this._tokenDecimals = new _NullableTokenDecimalMap.default();
    /**
     * @private
     * @type {HbarTransferMap}
     */

    this._hbarTransfers = new _HbarTransferMap.default();
    /**
     * @private
     * @type {TokenNftTransferMap}
     */

    this._nftTransfers = new _TokenNftTransferMap.default();
    this.setMaxTransactionFee(new _Hbar.default(1));

    for (const transfer of props.tokenTransfers != null ? props.tokenTransfers : []) {
      this.addTokenTransfer(transfer.tokenId, transfer.accountId, transfer.amount);
    }

    for (const transfer of props.hbarTransfers != null ? props.hbarTransfers : []) {
      this.addHbarTransfer(transfer.accountId, transfer.amount);
    }

    for (const transfer of props.nftTransfers != null ? props.nftTransfers : []) {
      this.addNftTransfer(transfer.tokenId, transfer.serial, transfer.sender, transfer.recipient);
    }
  }
  /**
   * @internal
   * @param {proto.ITransaction[]} transactions
   * @param {proto.ISignedTransaction[]} signedTransactions
   * @param {TransactionId[]} transactionIds
   * @param {AccountId[]} nodeIds
   * @param {proto.ITransactionBody[]} bodies
   * @returns {TransferTransaction}
   */


  static _fromProtobuf(transactions, signedTransactions, transactionIds, nodeIds, bodies) {
    const body = bodies[0];
    const cryptoTransfer =
    /** @type {proto.ICryptoTransferTransactionBody} */
    body.cryptoTransfer;
    const transfers = new TransferTransaction();

    for (const list of cryptoTransfer.tokenTransfers != null ? cryptoTransfer.tokenTransfers : []) {
      const tokenId = _TokenId.default._fromProtobuf(
      /** @type {proto.ITokenID} */
      list.token);

      for (const transfer of list.transfers != null ? list.transfers : []) {
        transfers.addTokenTransfer(tokenId, _AccountId.default._fromProtobuf(
        /** @type {proto.IAccountID} */
        transfer.accountID),
        /** @type {Long} */
        transfer.amount);
      }

      for (const transfer of list.nftTransfers != null ? list.nftTransfers : []) {
        transfers.addNftTransfer(tokenId,
        /** @type {Long} */
        transfer.serialNumber, _AccountId.default._fromProtobuf(
        /** @type {proto.IAccountID} */
        transfer.senderAccountID), _AccountId.default._fromProtobuf(
        /** @type {proto.IAccountID} */
        transfer.receiverAccountID));
      }
    }

    const accountAmounts = cryptoTransfer.transfers != null ? cryptoTransfer.transfers.accountAmounts != null ? cryptoTransfer.transfers.accountAmounts : [] : [];

    for (const aa of accountAmounts) {
      transfers.addHbarTransfer(_AccountId.default._fromProtobuf(
      /** @type {proto.IAccountID} */
      aa.accountID), _Hbar.default.fromTinybars(
      /** @type {Long} */
      aa.amount));
    }

    return _Transaction.default._fromProtobufTransactions(transfers, transactions, signedTransactions, transactionIds, nodeIds, bodies);
  }
  /**
   * @returns {TokenTransferMap}
   */


  get tokenTransfers() {
    return this._tokenTransfers;
  }
  /**
   * @param {TokenId | string} tokenId
   * @param {AccountId | string} accountId
   * @param {number | Long} amount
   * @returns {this}
   */


  addTokenTransfer(tokenId, accountId, amount) {
    this._requireNotFrozen();

    const token = tokenId instanceof _TokenId.default ? tokenId : _TokenId.default.fromString(tokenId);

    this._tokenTransfers.__set(token, accountId instanceof _AccountId.default ? accountId : _AccountId.default.fromString(accountId), amount instanceof _long.default ? amount : _long.default.fromNumber(amount));

    const currentDecimals = this._tokenDecimals.get(token);

    if (currentDecimals == null) {
      this._tokenDecimals._set(token, null);
    }

    return this;
  }
  /**
   * @param {TokenId | string} tokenId
   * @param {AccountId | string} accountId
   * @param {number | Long} amount
   * @param {number} decimals
   * @returns {this}
   */


  addTokenTransferWithDecimals(tokenId, accountId, amount, decimals) {
    this._requireNotFrozen();

    const token = tokenId instanceof _TokenId.default ? tokenId : _TokenId.default.fromString(tokenId);

    this._tokenTransfers.__set(token, accountId instanceof _AccountId.default ? accountId : _AccountId.default.fromString(accountId), amount instanceof _long.default ? amount : _long.default.fromNumber(amount));

    const currentDecimals = this._tokenDecimals.get(token);

    if (currentDecimals != null && currentDecimals != decimals) {
      throw new Error("token ID decimal mismatch");
    }

    if (currentDecimals == null) {
      this._tokenDecimals._set(token, decimals);
    }

    return this;
  }
  /**
   * @returns {NullableTokenDecimalMap}
   */


  get tokenIdDecimals() {
    return this._tokenDecimals;
  }
  /**
   * @returns {HbarTransferMap}
   */


  get hbarTransfers() {
    return this._hbarTransfers;
  }
  /**
   * @internal
   * @param {AccountId | string} accountId
   * @param {number | string | Long | LongObject | BigNumber | Hbar} amount
   * @returns {TransferTransaction}
   */


  addHbarTransfer(accountId, amount) {
    this._requireNotFrozen();

    if (this.hbarTransfers.get(accountId) != null) {
      amount = this._squashTransfers(accountId, amount);
    }

    this._hbarTransfers._set(accountId instanceof _AccountId.default ? accountId : _AccountId.default.fromString(accountId), amount instanceof _Hbar.default ? amount : new _Hbar.default(amount));

    return this;
  }
  /**
   * @internal
   * @param {AccountId | string} accountId
   * @param {number | string | Long | LongObject | BigNumber | Hbar} amount
   * @returns {Hbar}
   */


  _squashTransfers(accountId, amount) {
    let currentValue = this.hbarTransfers.get(accountId);
    return _Hbar.default.fromTinybars((currentValue == null ? _long.default.ZERO : currentValue.toTinybars()).add(amount instanceof _Hbar.default ? amount.toTinybars() : new _Hbar.default(amount).toTinybars()));
  }
  /**
   * @param {Client} client
   */


  _validateChecksums(client) {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    for (const [a, _] of this._hbarTransfers) {
      if (a != null) {
        a.validateChecksum(client);
      }
    }

    for (const [tokenId, transfers] of this._tokenTransfers) {
      if (tokenId != null) {
        tokenId.validateChecksum(client);
      } // eslint-disable-next-line @typescript-eslint/no-unused-vars


      for (const [a, _] of transfers) {
        if (a != null) {
          a.validateChecksum(client);
        }
      }
    }
  }
  /**
   * @returns {TokenNftTransferMap}
   */


  get nftTransfers() {
    return this._nftTransfers;
  }
  /**
   * @internal
   * @param {NftId | TokenId | string} tokenIdOrNftId
   * @param {AccountId | string | Long | number} senderAccountIdOrSerialNumber
   * @param {AccountId | string} recipientAccountIdOrSenderAccountId
   * @param {(AccountId | string)=} recipient
   * @returns {TransferTransaction}
   */


  addNftTransfer(tokenIdOrNftId, senderAccountIdOrSerialNumber, recipientAccountIdOrSenderAccountId, recipient) {
    this._requireNotFrozen();

    let tokenId;
    let serial;
    let senderId;
    let recipientId;

    if (typeof tokenIdOrNftId === "string") {
      if (tokenIdOrNftId.includes("/") || tokenIdOrNftId.includes("@")) {
        tokenIdOrNftId = _NftId.default.fromString(tokenIdOrNftId);
      } else {
        tokenIdOrNftId = _TokenId.default.fromString(tokenIdOrNftId);
      }
    }

    if (tokenIdOrNftId instanceof _NftId.default) {
      tokenId = tokenIdOrNftId.tokenId;
      serial = tokenIdOrNftId.serial;
      senderId =
      /** @type {AccountId | string} */
      senderAccountIdOrSerialNumber;
      recipientId =
      /** @type {AccountId | string} */
      recipientAccountIdOrSenderAccountId;
    } else if (tokenIdOrNftId instanceof _TokenId.default) {
      tokenId =
      /** @type {TokenId} */
      tokenIdOrNftId;
      serial =
      /** @type {Long|number} */
      senderAccountIdOrSerialNumber;
      senderId =
      /** @type {AccountId | string} */
      recipientAccountIdOrSenderAccountId;
      util.requireNonNull(recipient);
      recipientId =
      /** @type {AccountId | string} */
      recipient;
    } else {
      throw new Error("unintended type for tokenIdOrNftId");
    }

    this._nftTransfers.__set(typeof tokenId === "string" ? _TokenId.default.fromString(tokenId) : tokenId, {
      serial: typeof serial === "number" ? _long.default.fromNumber(serial) : serial,
      sender: typeof senderId === "string" ? _AccountId.default.fromString(senderId) : senderId,
      recipient: typeof recipientId === "string" ? _AccountId.default.fromString(recipientId) : recipientId
    });

    return this;
  }
  /**
   * @override
   * @internal
   * @param {Channel} channel
   * @param {proto.ITransaction} request
   * @returns {Promise<proto.ITransactionResponse>}
   */


  _execute(channel, request) {
    return channel.crypto.cryptoTransfer(request);
  }
  /**
   * @override
   * @protected
   * @returns {NonNullable<proto.TransactionBody["data"]>}
   */


  _getTransactionDataCase() {
    return "cryptoTransfer";
  }
  /**
   * @override
   * @protected
   * @returns {proto.ICryptoTransferTransactionBody}
   */


  _makeTransactionData() {
    /** @type {Set<string>} */
    const tokenIds = new Set();

    for (const tokenId of this._tokenTransfers.keys()) {
      tokenIds.add(tokenId.toString());
    }

    for (const tokenId of this._nftTransfers.keys()) {
      tokenIds.add(tokenId.toString());
    }

    const tokenTransfers = Array.from(tokenIds, tokenId => {
      const thisTokenTransfers = this._tokenTransfers.get(tokenId);

      const thisNftTransfers = this._nftTransfers.get(tokenId);

      let transfers;
      let nftTransfers;

      if (thisTokenTransfers != null) {
        transfers = Array.from(thisTokenTransfers, ([accountId, amount]) => {
          return {
            accountId,
            amount
          };
        });
        transfers.sort((a, b) => {
          const accountIdComparison = a.accountId.compare(b.accountId);

          if (accountIdComparison != 0) {
            return accountIdComparison;
          }

          return a.amount.compare(b.amount);
        });
      }

      if (thisNftTransfers != null) {
        // eslint-disable-next-line ie11/no-loop-func
        nftTransfers = thisNftTransfers.map(transfer => {
          return {
            sender: transfer.sender,
            recipient: transfer.recipient,
            serialNumber: transfer.serial
          };
        });
        nftTransfers.sort((a, b) => {
          const senderComparision = a.sender.compare(b.sender);

          if (senderComparision != 0) {
            return senderComparision;
          }

          const recipientComparision = a.recipient.compare(b.recipient);

          if (recipientComparision != 0) {
            return recipientComparision;
          }

          return a.serialNumber.compare(b.serialNumber);
        });
      }

      return {
        token: _TokenId.default.fromString(tokenId),
        transfers,
        nftTransfers
      };
    });
    const hbarTransfers = Array.from(this._hbarTransfers, ([accountId, amount]) => {
      return {
        accountId,
        amount
      };
    });
    tokenTransfers.sort((a, b) => {
      return a.token.compare(b.token);
    });
    hbarTransfers.sort((a, b) => {
      const accountIdComparison = a.accountId.compare(b.accountId);

      if (accountIdComparison != 0) {
        return accountIdComparison;
      }

      return a.amount.toTinybars().compare(b.amount.toTinybars());
    });
    return {
      transfers: {
        accountAmounts: hbarTransfers.map(transfer => {
          return {
            accountID: transfer.accountId._toProtobuf(),
            amount: transfer.amount.toTinybars()
          };
        })
      },
      tokenTransfers: tokenTransfers.map(tokenTransfer => {
        const expectedDecimals = this._tokenDecimals.get(tokenTransfer.token);

        return {
          token: tokenTransfer.token._toProtobuf(),
          expectedDecimals: expectedDecimals != null ? {
            value: expectedDecimals
          } : null,
          transfers: (tokenTransfer.transfers != null ? tokenTransfer.transfers : []).map(transfer => {
            return {
              accountID: transfer.accountId._toProtobuf(),
              amount: transfer.amount
            };
          }),
          nftTransfers: (tokenTransfer.nftTransfers != null ? tokenTransfer.nftTransfers : []).map(nftTransfer => {
            return {
              senderAccountID: nftTransfer.sender._toProtobuf(),
              receiverAccountID: nftTransfer.recipient._toProtobuf(),
              serialNumber: nftTransfer.serialNumber
            };
          })
        };
      })
    };
  }

}

exports.default = TransferTransaction;

_Transaction.TRANSACTION_REGISTRY.set("cryptoTransfer", // eslint-disable-next-line @typescript-eslint/unbound-method
TransferTransaction._fromProtobuf);
