// @hashgraph/sdk/lib/contract/DelegateContractId.cjs
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _Cache = _interopRequireDefault(require("../Cache.cjs"));

var _ContractId = _interopRequireDefault(require("./ContractId.cjs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @namespace {proto}
 * @typedef {import("@hashgraph/proto").IContractID} proto.IContractID
 * @typedef {import("@hashgraph/proto").IKey} proto.IKey
 */

/**
 * @typedef {import("long").Long} Long
 * @typedef {import("../client/Client.js").default<*, *>} Client
 */
class DelegateContractId extends _ContractId.default {
  /**
   * @param {number | Long | import("../EntityIdHelper").IEntityId} props
   * @param {(number | Long)=} realm
   * @param {(number | Long)=} num
   */
  constructor(props, realm, num) {
    super(props, realm, num);
  }
  /**
   * @param {string} text
   * @returns {DelegateContractId}
   */


  static fromString(text) {
    return new DelegateContractId(_ContractId.default.fromString(text));
  }
  /**
   * @internal
   * @param {proto.IContractID} id
   * @returns {DelegateContractId}
   */


  static _fromProtobuf(id) {
    return new DelegateContractId(_ContractId.default._fromProtobuf(id));
  }
  /**
   * @param {Uint8Array} bytes
   * @returns {DelegateContractId}
   */


  static fromBytes(bytes) {
    return new DelegateContractId(_ContractId.default.fromBytes(bytes));
  }
  /**
   * @param {string} address
   * @returns {DelegateContractId}
   */


  static fromSolidityAddress(address) {
    return new DelegateContractId(_ContractId.default.fromSolidityAddress(address));
  }
  /**
   * @returns {DelegateContractId}
   */


  clone() {
    const id = new DelegateContractId(this);
    id._checksum = this._checksum;
    return id;
  }
  /**
   * @returns {proto.IKey}
   */


  _toProtobufKey() {
    return {
      delegatableContractId: this._toProtobuf()
    };
  }
  /**
   * @param {proto.IContractID} key
   * @returns {DelegateContractId}
   */


  static __fromProtobufKey(key) {
    return DelegateContractId._fromProtobuf(key);
  }

}

exports.default = DelegateContractId;

_Cache.default.delegateContractId = key => DelegateContractId.__fromProtobufKey(key);
